ISSUE#12: Use-pen_state-in-direction-handling first

GijsterHaar opened the pull request:
This commit handles the basic distinction between PEN UP or PEN DOWN
When the state of the pen is PEN UP we can only move the pointer.
When the state of the pen is PEN DOWN the pointer will draw a line

GijsterHaar committed:
ISSUE#12: Use-pen_state-in-direction-handling first

GijsterHaar committed:
ISSUE#12: Use-pen_state-in-direction-handling_with_testing first

the-riviera-kid commented on tortoise_interpreter.py:
```diff
@@ -20,41 +20,54 @@ def sanitize_program(program):
-    # Comment
-    return [f'{i}. {COMMANDS[command[0]](command)}' for i, command in enumerate(program, 1)]
+    command_to_parse = []
+    pen_state = ''
```
So. I think that passing a state object from command to command is actually a really good idea :+1: However, why a string? If it was an object you'd be able to store more than one thing, give those items sensible names, etc etc.

the-riviera-kid commented on tortoise_interpreter.py:
```diff
@@ -20,41 +20,54 @@ def sanitize_program(program):
+    pen_state = ''
+    for i, command in enumerate(program, 1):
+        if command[0] in ['U', 'D']:
+            pen_state = get_pen_state(command)
```
I don't think that the `navigate_command_functions` function should need to know which commands alter the state. The Up/Down functions should probably do that themselves.

the-riviera-kid commented on tortoise_interpreter.py:
```diff
@@ -20,41 +20,54 @@ def sanitize_program(program):
-def format_direction_and_units(command, direction):
+def format_direction_and_units(command, pen_state, direction):
+    if pen_state == 'DOWN':
+        return f"Draw a line {command[1]} {format_string_for_singular_or_plural(command)} to the {direction}."
```
There are two lines in this function that are almost exactly the same. How could you reduce the duplication here?

GijsterHaar replied:
```
def format_direction_and_units(command, pen_state, direction):
        action = 'Move' if pen_state == 'UP' else 'Draw a line'
        return f"{action} {command[1]} {format_string_for_singular_or_plural(command)} to the {direction}."
```

the-riviera-kid reviewed (changes requested):
There's more to do, but I'm quite impressed; I thought we might need a few reviews before we got to this point. :+1: :tada:

GijsterHaar committed:
ISSUE#12: Use-pen_state-in-direction-handling first partly refactored

GijsterHaar committed:
ISSUE#12: Use-pen_state-in-direction-handling-second-refactor

Separated the get_pen_state function from the navigate_command_functions.
Added tests for the get_pen_state and the pen_up ignored.
Modified the test_navigate_command_functions test
Deleted the pen_up and pen_colour functions in the scope of
the test_navigate_command_functions test as they are not needed.

the-riviera-kid commented on example-program.tortoise:
```diff
@@ -1,7 +1,13 @@
-E 2  # then east 2
-S 1  # then back south
 U    # pen up
+U
```
Let's not alter `example.tortoise`. That's the only example from the book; let's keep that as-is.

the-riviera-kid commented on test_tortoise_interpreter.py:
```diff
@@ -4,7 +4,12 @@
+    pen_colour,
+    get_pen_state,
+    move_east,
+    move_north
```
Why have we imported these functions? Their behavior could already be tested through `main_tortoise`.

GijsterHaar replied:
after rewriting the test_navigate_command_functions() test, we do not need them anymore, deleted

the-riviera-kid commented on test_tortoise_interpreter.py:
```diff
@@ -58,22 +67,28 @@ def test_sanitized_programs(program, expected):
-    command_to_parse = navigate_command_functions(COMMANDS, program)
 
-    assert command_to_parse == ['1. P 3', '2. PEN UP']
+    COMMANDS = {'U': pen_up,'D': pen_down ,'P': pen_colour, 'N': move_north, 'E': move_east}
```
We shouldn't use the real handler functions here. This ties our test to an implementation detail, and is unnecessary. Implementing test functions shows that the function really is using the commands that are passed in; there's no hard-coding going on.

the-riviera-kid commented on tortoise_interpreter.py:
```diff
@@ -20,41 +19,59 @@ def sanitize_program(program):
+    command_to_parse = []
+    pen_state = ''
+    for i, command in enumerate(program, 1):
+        new_pen_state = get_pen_state(command, pen_state)
```
The aim was to move the knowledge about when to alter the state into the handler functions. This hasn't been achieved. The handler function should either alter the state it's given, or return a new state.

the-riviera-kid commented on tortoise_interpreter.py:
```diff
@@ -20,41 +19,59 @@ def sanitize_program(program):
+def pen_up(command, pen_state, new_pen_state):
+    if new_pen_state != pen_state:
+        return f'PEN UP'
+    return f'PEN UP (ignored)'
```
Is there a simpler way to write this? One that doesn't require the "PEN UP" string to be there twice?

the-riviera-kid commented on test_tortoise_interpreter.py:
```diff
@@ -58,22 +67,28 @@ def test_sanitized_programs(program, expected):
 
 def test_pen_up():
-    assert pen_up(['U']) == 'PEN UP'
+    assert pen_up(['U'], 'UP', 'DOWN') == 'PEN UP'
```
I know that `test_pen_up` was already here. However, I don't think it's a good idea to test low level details. You don't have to take them out, but maybe don't add any more.

GijsterHaar replied:
removed

the-riviera-kid commented on tortoise_interpreter.py:
```diff
@@ -20,41 +19,59 @@ def sanitize_program(program):
-    # Comment
-    return [f'{i}. {COMMANDS[command[0]](command)}' for i, command in enumerate(program, 1)]
+    command_to_parse = []
+    pen_state = ''
```
Should this default to empty? Is there a sensible default you could use? One that's implied by the ticket? :innocent:

GijsterHaar replied:
done

the-riviera-kid reviewed (changes requested):
There's thought gone into there, so it pains me to ask for so many changes. However, learning!

Do not shy about asking me questions; trying to work it out yourself is good, getting stuck is not. :+1:

GijsterHaar committed:
ISSUE#12: Added Pen State Up or Down

GijsterHaar committed:
ISSUE#12:  Use-pen_state-in-direction-handling - refactored

the-riviera-kid commented on test_tortoise_interpreter.py:
```diff
@@ -67,28 +61,18 @@ def test_sanitized_programs(program, expected):
+        return '___', pen_state
+    def double_colon(command, pen_state):
+        return '::', pen_state
+    def hard_stop(commands, pen_state):
```
I would like to see one of these handler functions use the arguments. Maybe the command should be `['H', '4']`, meaning it should print out four full stops. Maybe one could print out the pen state. 

However, this is much, much better than before :+1:

the-riviera-kid reviewed (changes requested):
You still need to address my comment about storing state as an object, rather than a string. :+1: 

However, it's pretty good at this point.

GijsterHaar committed:
ISSUE#12: Use-pen_state-in-direction-handling-fourth-refactor

the-riviera-kid commented on test_tortoise_interpreter.py:
```diff
@@ -62,17 +62,19 @@ def test_sanitized_programs(program, expected):
+        return ('___', pen_state), pen_state
     def double_colon(command, pen_state):
         return '::', pen_state
     def hard_stop(commands, pen_state):
```
This is good!

the-riviera-kid commented on test_tortoise_interpreter.py:
```diff
@@ -62,17 +62,19 @@ def test_sanitized_programs(program, expected):
 def test_navigate_command_functions():
     def underscore(command, pen_state):
-        return '___', pen_state
+        pen_state = "UP"
```
This doesn't use the `pen_state` argument. It uses the _name_ "pen_state", but it never reads the argument it's given. 

The reason I want the test handlers to use the `command` and `pen_state` arguments is to prove that `navigate_command_functions` is passing in the correct information.

the-riviera-kid commented on tortoise_interpreter.py:
```diff
@@ -21,49 +21,49 @@ def sanitize_program(program):
 def navigate_command_functions(COMMANDS, program):
     command_to_parse = []
-    pen_state = 'UP'
+    draw = False
```
So, this used to be `pen_state`, with a string value. The advice was to change the string to an object, so that we could store more information and name it well.

Now, we have `draw` which stores a boolean. This stores **_less_** information, and isn't as well named. We should have a talk about this one, because I think there's some miscommunication going on.

GijsterHaar committed:
ISSUE#12: Use-pen_state-in-direction-handling further refactor

Pen_state in navigate_command_functions test

GijsterHaar committed:
ISSUE#!12: Use_pen_state_in_direction_handling_pen_state_is_object

the-riviera-kid committed:
Merge branch 'main' into issue#12
